File: crates/cmux-workspaces/src/placement.rs
//! Port of the pure `insertionIndex` arithmetic from
//! app `Sources/WorkspacePlacement+Resolution.swift`.
//!
//! `effectivePlacement` (settings reads) is intentionally left to host wiring;
//! only the pure index arithmetic is ported here, reusing
//! [`cmux_config::NewWorkspacePlacement`].
//!
//! DIVERGENCE: Swift uses `Int` throughout; the port uses `i64` to mirror the
//! signed clamping (`max(0, min(...))`) exactly.

use cmux_config::NewWorkspacePlacement;

/// The insertion index for a new workspace under `placement`, given the current
/// selection and pinned-prefix shape of the tab list. Pure arithmetic; clamps
/// every input into the valid range.
pub fn insertion_index(
    placement: NewWorkspacePlacement,
    selected_index: Option<i64>,
    selected_is_pinned: bool,
    pinned_count: i64,
    total_count: i64,
) -> i64 {
    let clamped_total_count = total_count.max(0);
    let clamped_pinned_count = pinned_count.min(clamped_total_count).max(0);

    match placement {
        // Keep pinned workspaces grouped at the top by inserting ahead of
        // unpinned items.
        NewWorkspacePlacement::Top => clamped_pinned_count,
        NewWorkspacePlacement::End => clamped_total_count,
        NewWorkspacePlacement::AfterCurrent => {
            let Some(selected_index) = selected_index else {
                return clamped_total_count;
            };
            if clamped_total_count == 0 {
                return clamped_total_count;
            }
            let clamped_selected_index = selected_index.min(clamped_total_count - 1).max(0);
            if selected_is_pinned {
                return clamped_pinned_count;
            }
            (clamped_selected_index + 1).min(clamped_total_count)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn top_inserts_after_pinned_prefix() {
        assert_eq!(
            insertion_index(NewWorkspacePlacement::Top, None, false, 2, 5),
            2
        );
        // pinnedCount clamps into [0, totalCount].
        assert_eq!(
            insertion_index(NewWorkspacePlacement::Top, None, false, 9, 3),
            3
        );
        assert_eq!(
            insertion_index(NewWorkspacePlacement::Top, None, false, -1, 3),
            0
        );
    }

    #[test]
    fn end_inserts_at_total() {
        assert_eq!(
            insertion_index(NewWorkspacePlacement::End, Some(0), false, 1, 4),
            4
        );
        // Negative total clamps to 0.
        assert_eq!(
            insertion_index(NewWorkspacePlacement::End, None, false, 0, -2),
            0
        );
    }

    #[test]
    fn after_current_places_after_selected_unpinned() {
        // selected at index 1 (unpinned) → insert at 2.
        assert_eq!(
            insertion_index(NewWorkspacePlacement::AfterCurrent, Some(1), false, 1, 5),
            2
        );
    }

    #[test]
    fn after_current_pinned_selection_inserts_at_pinned_boundary() {
        // A pinned selection inserts at the pinned boundary, not after itself.
        assert_eq!(
            insertion_index(NewWorkspacePlacement::AfterCurrent, Some(0), true, 2, 5),
            2
        );
    }

    #[test]
    fn after_current_without_selection_falls_back_to_end() {
        assert_eq!(
            insertion_index(NewWorkspacePlacement::AfterCurrent, None, false, 1, 5),
            5
        );
        // Empty list → 0 even with a (stale) selected index.
        assert_eq!(
            insertion_index(NewWorkspacePlacement::AfterCurrent, Some(3), false, 0, 0),
            0
        );
    }

    #[test]
    fn after_current_clamps_selected_index_into_range() {
        // Out-of-range selectedIndex clamps to totalCount-1, then +1 clamps to total.
        assert_eq!(
            insertion_index(NewWorkspacePlacement::AfterCurrent, Some(99), false, 0, 3),
            3
        );
        // Negative selectedIndex clamps to 0 → insert at 1.
        assert_eq!(
            insertion_index(NewWorkspacePlacement::AfterCurrent, Some(-4), false, 0, 3),
            1
        );
    }
}

File: apps/desktop/web/src/sidebar/placement.ts
// New-workspace sidebar placement — port of:
// - `Sources/WorkspacePlacement+Resolution.swift:13-59`
//   (`effectivePlacement` 17-29, `insertionIndex` 34-59 — pure arithmetic,
//   staged in Swift for extraction into CmuxWorkspaces),
// - the `WorkspacePlacement` enum,
//   `Packages/macOS/CmuxSettings/Sources/CmuxSettings/Values/WorkspacePlacement.swift:4-6`,
// - the composing fold `TabManager.newTabInsertIndex`,
//   `Sources/TabManager.swift:1473-1506`.
//
// This is the SIDEBAR-level placement for `app.newWorkspacePlacement`. The
// group-level placement (`WorkspaceGroupNewPlacement` + `placeWithinGroup`) is
// a different feature and deliberately NOT part of this module.
//
// All indices are small integers; like the Swift Ints, callers are trusted to
// pass integers (no silent float truncation is performed).

/**
 * Where a newly-created workspace lands in the sidebar. Exact Swift raw
 * values (WorkspacePlacement.swift:4-6) — note camelCase `afterCurrent`.
 */
export type WorkspacePlacement = "top" | "end" | "afterCurrent";

/**
 * Strict exact-match parse of a persisted raw value, mirroring Swift's plain
 * `WorkspacePlacement(rawValue:)` decode (invalid values are logged-and-
 * skipped by the settings store; catalog default is `"afterCurrent"`).
 *
 * Deliberately NOT the tolerant multi-spelling parse — that belongs to the
 * separate group-placement setting (`WorkspaceGroupNewPlacement`).
 */
export function parseWorkspacePlacement(raw: string): WorkspacePlacement | null {
  switch (raw) {
    case "top":
    case "end":
    case "afterCurrent":
      return raw;
    default:
      return null;
  }
}

/**
 * The current selection + pinned-prefix shape of the tab list, as consumed by
 * `insertionIndex`. Mirrors the Swift parameter list
 * (WorkspacePlacement+Resolution.swift:34-38).
 */
export interface InsertionContext {
  selectedIndex: number | null;
  selectedIsPinned: boolean;
  pinnedCount: number;
  totalCount: number;
}

/**
 * The insertion index for a new workspace under `placement`. Mirrors
 * `WorkspacePlacement.insertionIndex` (WorkspacePlacement+Resolution.swift:34-59)
 * exactly: pure arithmetic, every input clamped into the valid range —
 * negative totals, `pinnedCount > totalCount`, and out-of-range
 * `selectedIndex` never throw.
 */
export function insertionIndex(
  placement: WorkspacePlacement,
  { selectedIndex, selectedIsPinned, pinnedCount, totalCount }: InsertionContext,
): number {
  const clampedTotalCount = Math.max(0, totalCount);
  const clampedPinnedCount = Math.max(0, Math.min(pinnedCount, clampedTotalCount));

  switch (placement) {
    case "top":
      // Keep pinned workspaces grouped at the top by inserting ahead of unpinned items.
      return clampedPinnedCount;
    case "end":
      return clampedTotalCount;
    case "afterCurrent": {
      if (selectedIndex === null || clampedTotalCount === 0) {
        return clampedTotalCount;
      }
      const clampedSelectedIndex = Math.max(0, Math.min(selectedIndex, clampedTotalCount - 1));
      if (selectedIsPinned) {
        return clampedPinnedCount;
      }
      return Math.min(clampedSelectedIndex + 1, clampedTotalCount);
    }
  }
}

/** One tab in the already-resolved current sidebar order. */
export interface PlacementTab {
  id: string;
  isPinned: boolean;
}

/**
 * Folds a workspace-creation snapshot into `insertionIndex`. Mirrors
 * `TabManager.newTabInsertIndex` (TabManager.swift:1489-1505).
 *
 * The caller supplies the current ordered tab list (Swift's
 * `orderedLiveWorkspaceCreationTabs(from:) ?? snapshot.tabs` live-refresh is
 * host-side and already resolved here).
 *
 * `pinnedCount` counts pinned tabs across the WHOLE list, not just the
 * prefix — a pinned tab out of prefix position still increments the count,
 * exactly like the Swift reduce.
 */
export function newTabInsertIndex(
  placement: WorkspacePlacement,
  tabs: readonly PlacementTab[],
  selectedTabId: string | null,
  selectedTabWasPinned: boolean,
): number {
  const pinnedCount = tabs.reduce((partial, tab) => (tab.isPinned ? partial + 1 : partial), 0);

  switch (placement) {
    case "top":
      return pinnedCount;
    case "end":
      return tabs.length;
    case "afterCurrent": {
      if (selectedTabId !== null) {
        const selectedIndex = tabs.findIndex((tab) => tab.id === selectedTabId);
        if (selectedIndex !== -1) {
          return insertionIndex("afterCurrent", {
            selectedIndex,
            selectedIsPinned: selectedTabWasPinned,
            pinnedCount,
            totalCount: tabs.length,
          });
        }
      }
      return selectedTabWasPinned ? pinnedCount : tabs.length;
    }
  }
}

/**
 * The placement to apply for a new workspace. Mirrors
 * `WorkspacePlacement.effectivePlacement`
 * (WorkspacePlacement+Resolution.swift:17-29): an explicit call-site override
 * wins, then iMessage mode pins `"top"`, then the stored
 * `app.newWorkspacePlacement` setting (catalog default `"afterCurrent"`,
 * supplied by the web settings layer).
 *
 * iMessage mode is a macOS feature; web callers pass `false`, but the
 * parameter is kept for byte-level rule parity.
 */
export function effectivePlacement(
  placementOverride: WorkspacePlacement | null,
  iMessageMode: boolean,
  stored: WorkspacePlacement,
): WorkspacePlacement {
  if (placementOverride !== null) {
    return placementOverride;
  }
  if (iMessageMode) {
    return "top";
  }
  return stored;
}
